lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,429 @@
+pub struct ImmutableIdTracker {
+    #[serde(skip)]
+    path: PathBuf,
+
+    deleted: BitVec,
+    dirty: Arc<RwLock<bool>>,
+
+    mappings: PointMappings,
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
+pub(super) struct PointMappings {
+    pub(crate) internal_to_version: Vec<SeqNumberType>,
+
+    pub(crate) internal_to_external: Vec<StoredPointId>,
+    pub(crate) external_to_internal: BTreeMap<StoredPointId, PointOffsetType>,
+}
+
+impl ImmutableIdTracker {
+    pub fn open(segment_path: &Path) -> OperationResult<Self> {
+        let deleted: BitVec = Self::open_file(segment_path.join(DELETED_FILE_NAME).as_path())?;
+        let mappings: PointMappings =
+            Self::open_file(segment_path.join(MAPPINGS_FILE_NAME).as_path())?;
+
+        Ok(Self {
+            path: segment_path.to_path_buf(),
+            deleted,
+            dirty: Arc::new(RwLock::new(false)),
+            mappings,
+        })
+    }
+
+    pub(super) fn new(path: PathBuf, deleted: BitVec, mappings: PointMappings) -> Self {
+        Self {
+            path,
+            deleted,
+            dirty: Arc::new(RwLock::new(false)),
+            mappings,
+        }
+    }
+
+    pub(super) fn save(&self) -> OperationResult<()> {
+        Self::save_file(&self.deleted_file_path(), &self.deleted)?;
+        Self::save_file(&self.mappings_file_path(), &self.mappings)?;
+        *self.dirty.write() = false;
+        Ok(())
+    }
+
+    fn open_file<T: DeserializeOwned>(file: &Path) -> OperationResult<T> {
+        let file = File::open(file)?;
+        let reader = BufReader::new(file);
+        bincode::deserialize_from(reader).map_err(|err| OperationError::InconsistentStorage {
+            description: format!("id tracker can't be deserialized: {:?}", err),
+        })
+    }
+
+    fn deleted_file_path(&self) -> PathBuf {
+        self.path.join(DELETED_FILE_NAME)
+    }
+
+    fn mappings_file_path(&self) -> PathBuf {
+        self.path.join(MAPPINGS_FILE_NAME)
+    }
+
+    pub(crate) fn save_file<T: Serialize>(path: &Path, value: &T) -> OperationResult<()> {
+        let file = File::create(path)?;
+        let writer = BufWriter::new(file);
+        bincode::serialize_into(writer, value).map_err(|err| {
+            OperationError::InconsistentStorage {
+                description: format!("id tracker can't be serialized: {:?}", err),
+            }
+        })?;
+        Ok(())
+    }
+}
+
+impl IdTracker for ImmutableIdTracker {
+    fn internal_version(&self, internal_id: PointOffsetType) -> Option<SeqNumberType> {
+        self.mappings
+            .internal_to_version
+            .get(internal_id as usize)
+            .copied()
+    }
+
+    fn set_internal_version(
+        &mut self,
+        _internal_id: PointOffsetType,
+        _version: SeqNumberType,
+    ) -> OperationResult<()> {
+        panic!("Trying to call a mutating function (`set_internal_version`) of an immutable id tracker");
```

Comment:
We will need this eventually. Or well, I believe we need to be able to use this to update the version on delete.